### PR TITLE
Log an error when constraining non-abstract type to specialization

### PR DESCRIPTION
**Pull request by ngfreiter**
A type that is not abstract cannot be constrained to any specializations of that type. When the type is not abstract, the type is final, and so it can only be constrained to profiles of that type. The best example is `Quantity`, which is not abstract, and has both specializations, such as `Duration`, and profiles, such as `SimpleQuantity`. So an element of type `Quantity`, say `valueQuantity`, can have the following rule:
```
* valueQuantity only SimpleQuantity
```
but cannot have:
```
* valueQuantity only Duration
```

SUSHI used to allow the second rule, but the IG Publisher does not, so it would have errors. Now the second rule is disallowed, and an error is logged. The justification for this is in https://chat.fhir.org/#narrow/stream/179177-conformance/topic/Narrow.20Quantity.20to.20Duration.20and.20Age. I don't know if there is a location in the spec that really dives into this distinction. The only example I can find of a non-abstract type being specialized is `Quantity`, but the code shouldn't be specific to `Quantity`.

### CODE REVIEW

**Comment by ngfreiter** [comment]
Fixes #447

File: test/export/StructureDefinitionExporter.test.ts [hunk 1]
```diff
@@ -1595,22 +1595,21 @@ describe('StructureDefinitionExporter', () => {
     // Quantity is the first ancestor of Duration and Age
 
     // * value[x] only Duration or Age
-    // * value[x].comparator = #>=
+    // * value[x].comparator = #>
```
Comment (mint-thompson) [inline test/export/StructureDefinitionExporter.test.ts:1598]: Very minor, but you can undo this change, as the comparator is still `>=`

File: test/export/StructureDefinitionExporter.test.ts [hunk 2]
```diff
@@ -1595,22 +1595,21 @@ describe('StructureDefinitionExporter', () => {
...
     fixedValueRule.fixedValue = new FshCode('>=');
-    profile.rules.push(fixedValueRule);
+    extension.rules.push(fixedValueRule);
 
-    exporter.exportStructDef(profile);
-    const sd = pkg.profiles[0];
-    const fixedElement = sd.findElement('Observation.value[x].comparator');
+    exporter.exportStructDef(extension);
+    const sd = pkg.extensions[0];
+    const fixedElement = sd.findElement('Extension.value[x].comparator');
```
Comment (mint-thompson) [inline test/export/StructureDefinitionExporter.test.ts:1612]: Just to make sure I understand: the type on `value[x]` would be `Quantity`. Would the type have profiles for `Age` and `Duration`, or would it have no profiles?

Comment (ngfreiter) [reply test/export/StructureDefinitionExporter.test.ts:1612]: In this case the type on `value[x]` does actually get set to `Duration` or `Age`. `Duration` and `Age` are specializations of `Quantity`, but they are treated just as if they were any other type, because `Quantity` is not an abstract type, so it cannot be replaced by either `Duration` or `Age`. So on the `Observation.value[x]`, you cannot constrain the type to `Duration`, because `Duration` is not one of the choices for that `value[x]` element. `Quantity` is, but we cannot replace `Quantity` with `Duration`. That is why I had to change this test to use `Extension`, since `Duration` and `Age` are both explicitly listed in the list of types the `Extension.value[x]` choice can take on.

Comment (mint-thompson) [reply test/export/StructureDefinitionExporter.test.ts:1612]: Ahh, okay. Thank you for clarifying!
